service/udp.go

```diff
@@ -79,130 +87,286 @@ func findAccessKeyUDP(clientIP netip.Addr, dst, src []byte, cipherList CipherLis
+		targetConnFactory: func() (net.PacketConn, error) {
+			pc, err := net.ListenPacket("udp", "")
+			if err != nil {
+				return nil, fmt.Errorf("failed to create UDP socket: %v", err)
+			}
+			return &timedPacketConn{
+				PacketConn:     pc,
+				defaultTimeout: natTimeout,
+			}, nil
+		},
 	}
 }
 
-// PacketHandler is a running UDP shadowsocks proxy that can be stopped.
-type PacketHandler interface {
+// AssociationHandler is a handler that handles UDP assocations.
+type AssociationHandler interface {
+	Handle(association net.Conn, metrics UDPAssocationMetrics)
 	// SetLogger sets the logger used to log messages. Uses a no-op logger if nil.
 	SetLogger(l *slog.Logger)
 	// SetTargetIPValidator sets the function to be used to validate the target IP addresses.
 	SetTargetIPValidator(targetIPValidator onet.TargetIPValidator)
-	// Handle returns after clientConn closes and all the sub goroutines return.
-	Handle(clientConn net.PacketConn)
+	// SetTargetConnFactory sets the function to be used to create new target connections.
+	SetTargetConnFactory(factory func() (net.PacketConn, error))
 }
 
-func (h *packetHandler) SetLogger(l *slog.Logger) {
+func (h *associationHandler) SetLogger(l *slog.Logger) {
 	if l == nil {
 		l = noopLogger()
 	}
 	h.logger = l
 }
 
-func (h *packetHandler) SetTargetIPValidator(targetIPValidator onet.TargetIPValidator) {
+func (h *associationHandler) SetTargetIPValidator(targetIPValidator onet.TargetIPValidator) {
 	h.targetIPValidator = targetIPValidator
 }
 
-// Listen on addr for encrypted packets and basically do UDP NAT.
-// We take the ciphers as a pointer because it gets replaced on config updates.
-func (h *packetHandler) Handle(clientConn net.PacketConn) {
-	nm := newNATmap(h.natTimeout, h.m, h.logger)
+func (h *associationHandler) SetTargetConnFactory(factory func() (net.PacketConn, error)) {
+	h.targetConnFactory = factory
+}
+
+type AssocationHandleFunc func(assocation net.Conn)
+
+// PacketServe listens for UDP packets on the provided [net.PacketConn], creates
+// creates and manages NAT associations, and invokes the provided `handle`
+// function for each association. It uses a NAT map to track active associations
+// and handles their lifecycle.
+func PacketServe(clientConn net.PacketConn, handle AssocationHandleFunc, metrics NATMetrics) {
+	// This goroutine continuously reads from clientConn and sends the received data
+	// to readCh. It uses a buffer pool (readBufPool) to efficiently manage buffers
+	// and minimize allocations. The LazySlice is sent along with the read event
+	// to allow the receiver to release the buffer back to the pool after processing.
+	readCh := make(chan readEvent, 10)
+	go func() {
+		for {
+			lazySlice := readBufPool.LazySlice()
+			buffer := lazySlice.Acquire()
+			n, addr, err := clientConn.ReadFrom(buffer)
+			if err != nil {
+				lazySlice.Release()
+				if errors.Is(err, net.ErrClosed) {
+					readCh <- readEvent{err: err}
```

Comment:
You could also simply close the readCh, which I think it clearer. You need to check for closed in the event loop.